compiler/circle-mpqsolver/src/bisection/BisectionSolver.cpp

```diff
@@ -0,0 +1,227 @@
+  VERBOSE(l, 0) << "int8_quantization_error " << int8_qerror << std::endl;
+
+  _qerror = int16_qerror + _qerror_ratio * std::fabs(int8_qerror - int16_qerror);
+  VERBOSE(l, 0) << "target quantization error " << _qerror << std::endl;
+
+  if (int16_qerror > _qerror)
+  {
+    std::cerr << "ERROR: minimal targeted is an error of Q16 quantized model" << std::endl;
+    return nullptr;
+  }
+  if (int8_qerror < _qerror)
+  {
+    // no need for bisectioning just return Q8 model
+    if (!_quantizer->quantize(module.get(), "uint8", layer_params))
+    {
+      std::cerr << "ERROR: failed to quantize model" << std::endl;
+      return nullptr;
+    }
+  }
+
+  int last_depth = -1;
+  float best_depth = -1;
+  LayerParams best_params;
+  auto graph = module->graph(0);
+  auto active_nodes = loco::active_nodes(loco::output_nodes(graph));
+  // input and output nodes are not valid for quantization, so let's remove them
+  for (auto node : loco::input_nodes(graph))
+  {
+    active_nodes.erase(node);
+  }
+  for (auto node : loco::output_nodes(graph))
+  {
+    active_nodes.erase(node);
+  }
+
+  // let's decide whether nodes at input are more suspectible to be quantized into Q16, than at
+  // output
+  bool int16_at_input = true;
+  if (auto option = options())
+  {
+    if (option->query(Options::Q16AtInput::True))
+    {
+      int16_at_input = true;
+    }
+    else if (option->query(Options::Q16AtInput::False))
+    {
+      int16_at_input = false;
+    }
+  }
+
+  while (true)
+  {
+    float cut_depth = 0.5f * (min_depth + max_depth);
+
+    if (last_depth == int(cut_depth))
+    {
+      break;
+    }
+    last_depth = int(cut_depth);
+
+    LayerParams layer_params;
+    for (auto &node : active_nodes)
+    {
+      auto cur_node = loco::must_cast<luci::CircleNode *>(node);
+      auto iter = nodes_depth.find(cur_node);
+      if (iter == nodes_depth.end())
+      {
+        continue;
+      }
+
+      float depth = iter->second;
+      if ((depth <= cut_depth && int16_at_input) || (depth >= cut_depth && !int16_at_input))
+      {
+        auto layer_param = std::make_shared<LayerParam>();
+        {
+          layer_param->name = cur_node->name();
+          layer_param->dtype = "int16";
+          layer_param->granularity = "channel";
+        }
+
+        layer_params.emplace_back(layer_param);
+      }
+    }
+
+    float cur_accuracy = evaluate(evaluator, module_path, "uint8", layer_params);
+    VERBOSE(l, 0) << cut_depth << " : " << cur_accuracy << std::endl;
+
+    if (cur_accuracy < _qerror)
+    {
+      int16_at_input ? (max_depth = cut_depth) : (min_depth = cut_depth);
+      best_params = layer_params;
+      best_depth = cut_depth;
+    }
+    else
+    {
+      int16_at_input ? (min_depth = cut_depth) : (max_depth = cut_depth);
+    }
+  }
+
+  if (best_params.empty())
+  {
+    std::cerr << "Failed to find any configuration" << std::endl;
+    return nullptr;
```

Comment:
When this happen? 

We already checked target qerror is higher than full q16 qerror.

So, at least full q16 would be valid configuration.

Reply:
It's a bit redundant. The final check is to make it more safe. I'll remove it.